### PR TITLE
Update configuration parameters in watchdog on instance restart

### PR DESCRIPTION
**Pull request by LeonidVas**

### CODE REVIEW

File: cli/running/watchdog.go [hunk 1]
```diff
@@ -10,6 +10,19 @@ import (
 	"github.com/tarantool/tt/cli/ttlog"
 )
 
+// Provider interface provides Watchdog methods to get objects whose creation
+// and updating may depend on changing external parameters (such as configuration
+// file).
+type Provider interface {
```
Comment (psergee) [inline cli/running/watchdog.go:16]: I don't have an idea for a good name yet, but "Provider" looks too common for me. and does not give any clue of what it is for.

File: cli/running/running.go [hunk 2]
```diff
@@ -42,6 +43,90 @@ type RunOpts struct {
...
+	args := []string{provider.cmdCtx.Running.AppName}
+	if err = FillCtx(cliOpts, provider.cmdCtx, args); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// createInstance reads config and creates an Instance.
+func (provider *providerImpl) CreateInstance(logger *ttlog.Logger) (*Instance, error) {
```
[the remark this answers is not in the document]
Comment (LeonidVas) [reply cli/running/running.go:67]: If I understand the question correctly, the answer is https://github.com/tarantool/tt/pull/115#discussion_r942416475

File: cli/running/watchdog.go (hunk 1 above)
Comment (LeonidVas) [reply cli/running/watchdog.go:16]: It is a facade for interaction with the outside world. I'm willing to settle for another name, but I couldn't think of a good enough one.

File: cli/running/watchdog_test.go [hunk 3]
```diff
@@ -21,6 +21,35 @@ const (
...
+	logger *ttlog.Logger
+	// dataDir used by an Instance.
+	dataDir string
+	// restartable indicates the need to restart the instance in case of a crash.
+	restartable bool
+}
+
+// createInstance reads config and creates an Instance.
+func (provider *providerTestImpl) CreateInstance(logger *ttlog.Logger) (*Instance, error) {
+	return NewInstance(provider.tarantool, provider.appPath, "", os.Environ(), logger, provider.dataDir)
```
Comment (better0fdead) [inline cli/running/watchdog_test.go:40]: `line is 104 characters (lll)`
Make line shorter (<100 characters)

File: cli/running/watchdog_test.go [hunk 4]
```diff
@@ -44,12 +73,12 @@ func createTestWatchdog(t *testing.T, restartable bool) *Watchdog {
 	assert.Nilf(err, `Can't find a tarantool binary. Error: "%v".`, err)
 
 	logger := ttlog.NewCustomLogger(io.Discard, "", 0)
-	inst, err := NewInstance(tarantoolBin, appPath, "", os.Environ(), logger, dataDir)
-	assert.Nilf(err, `Can't create an instance. Error: "%v".`, err)
 
-	wd := NewWatchdog(inst, restartable, wdTestRestartTimeout, logger)
+	provider := providerTestImpl{tarantool: tarantoolBin, appPath: appPath, logger: logger, dataDir: dataDir, restartable: restartable}
```
Comment (better0fdead) [inline cli/running/watchdog_test.go:77]: `line is 135 characters (lll)`
Make line shorter (<100 characters)

File: cli/ttlog/logger.go [hunk 5]
```diff
@@ -63,3 +65,17 @@ func (logger *Logger) Rotate() error {
 
 	return logger.ljLogger.Rotate()
 }
+
+// GetOpts return the parameters that were used to create the logger.
```
Comment (better0fdead) [inline cli/ttlog/logger.go:69]:
```suggestion
// GetOpts returns the parameters that were used to create the logger.
```

File: cli/running/watchdog_test.go [hunk 6]
```diff
@@ -21,6 +21,35 @@ const (
 	wdTestStopTimeout    = 2 * time.Second
 )
 
+// providerTestImpl is an implementation of the Watchdog provider used for tests.
+type providerTestImpl struct {
+	// tarantool is the tarantool binary in use.
+	tarantool string
+	// appPath the path to the application.
```
Comment (better0fdead) [inline cli/running/watchdog_test.go:28]:
```suggestion
	// appPath is the path to the application.
```

**Review by better0fdead — CHANGES_REQUESTED** [review]
Thanks for patch.
I'm ok just small changes + added test for new behaviour of watchdog.

File: cli/running/watchdog_test.go [hunk 7]
```diff
@@ -58,7 +58,7 @@ func cleanupTempDir(tempDir string) {
 }
 
 // createTestWatchdog creates an instance and a watchdog for the test.
-func createTestWatchdog(t *testing.T, restartable bool) *Watchdog {
+func createTestWatchdog(t *testing.T, restartable bool) (*Watchdog, *providerTestImpl) {
```
Comment (LeonidVas) [inline cli/running/watchdog_test.go:61]: Provider is present in watchdog why are you returning it additionally?

File: cli/running/watchdog_test.go [hunk 8]
```diff
@@ -157,3 +157,35 @@ func TestWatchdogNotRestartable(t *testing.T) {
 	case <-wdDoneChan:
 	}
 }
+
+func TestWatchdogRereadParams(t *testing.T) {
```
Comment (LeonidVas) [inline cli/running/watchdog_test.go:161]: I could be wrong, but something similar is tested in other tests. I would like to see an integration test that re-reads the configuration from the file.

**Review by better0fdead — APPROVED** [review]

File: cli/running/watchdog_test.go (hunk 7 above)
Comment (better0fdead) [reply cli/running/watchdog_test.go]: removed commit

File: cli/running/watchdog_test.go (hunk 8 above)
Comment (better0fdead) [reply cli/running/watchdog_test.go]: removed commit

File: test/integration/running/test_app/test_app_restart.lua [hunk 9]
```diff
@@ -0,0 +1,3 @@
+local fiber = require('fiber')
```
Comment (LeonidVas) [inline test/integration/running/test_app/test_app_restart.lua:1]: An instance of such application will be stopped when the code is executed.
It is not suitable for testing this functionality.

Comment (LeonidVas) [inline test/integration/running/test_app/test_app_restart.lua:1]: Where do you use fiber?

File: test/integration/running/test_app/test_app_restart.lua [hunk 10]
```diff
@@ -0,0 +1,3 @@
...
+
+print(a[7])
```
Comment (LeonidVas) [inline test/integration/running/test_app/test_app_restart.lua:3]: What is it for?

File: test/integration/running/test_running.py [hunk 11]
```diff
@@ -139,3 +142,50 @@ def test_logrotate(tt_cmd, tmpdir):
...
+
+    # Check status.
+    status_cmd = [tt_cmd, "status", random_name]
+    status_rc, status_out = run_command_and_get_output(status_cmd, cwd=tmpdir)
+    assert status_rc == 0
+    assert re.search(r"RUNNING. PID: \d+.", status_out)
+
+    with open(configPath, 'w') as f:
+        f.write('tt:\n  app:\n    restart_on_failure: false')
+    time.sleep(5)
```
Comment (LeonidVas) [inline test/integration/running/test_running.py:181]: As far as I understand it, you just wait for the instance to finish executing the code and then check to see if it has been restarted.
This is incorrect. I think the code should not be restarted if the instance was completed correctly. Sounds like a bug in watchdog, I will discuss this point with PMs.

Comment (LeonidVas) [reply test/integration/running/test_running.py:181]: I think we should discuss with a voice how best to stop the instance here.

File: test/integration/running/test_app/test_app_restart.lua (hunk 9 above)
Comment (better0fdead) [reply test/integration/running/test_app/test_app_restart.lua]: removed

File: test/integration/running/test_app/test_app_restart.lua (hunk 10 above)
Comment (better0fdead) [reply test/integration/running/test_app/test_app_restart.lua:2]: it causes an error, so instance fails

File: test/integration/running/test_running.py (hunk 11 above)
Comment (better0fdead) [reply test/integration/running/test_running.py:182]: disscused in a voice call

File: test/integration/running/test_app/test_app_restart.lua (hunk 10 above)
Comment (better0fdead) [reply test/integration/running/test_app/test_app_restart.lua:2]: it causes an error, so instance fails

**Comment by LeonidVas** [comment]
I merged all commits into master without integration test. Do a rebase.

File: test/integration/running/test_running.py [hunk 12]
```diff
@@ -139,3 +140,91 @@ def test_logrotate(tt_cmd, tmpdir):
...
+                if "started" in log:
+                    flag = 1
+        except:
+            continue
+    
+    # Kill instance child process.
+    kill_child_process(pid, tmpdir)
+
+    # Check status, it should be running, since instance restarts after failure.
+    status_cmd = [tt_cmd, "status", inst_name]
```
Comment (LeonidVas) [inline test/integration/running/test_running.py:196]: How do you check that the instance already restarted?

File: test/integration/running/test_running.py [hunk 13]
```diff
@@ -139,3 +140,91 @@ def test_logrotate(tt_cmd, tmpdir):
...
+    status_rc, status_out = run_command_and_get_output(status_cmd, cwd=tmpdir)
+    assert status_rc == 0
+    assert re.search(r"RUNNING. PID: \d+.", status_out)
+
+    # Create test config with restart_on_failure false.
+    with open(configPath, 'w') as f:
+        f.write("tt:\n  app:\n    restart_on_failure: false\n    log_dir: "+"str(tmpdir)+\n    log_maxsize: 10    \n    log_maxage: 1\n")
+    
+    # Wait for child process of instance to start.
```
Comment (LeonidVas) [inline test/integration/running/test_running.py:205]: You already write this twice. Move this functionality to a separate function.

File: test/integration/running/test_running.py [hunk 14]
```diff
@@ -139,3 +140,91 @@ def test_logrotate(tt_cmd, tmpdir):
...
+                    pass
+                    last_line = line
+                if "started" in last_line:
+                    flag = 1
+        except:
+            continue
+    
+    # Kill instance child process.
+    kill_child_process(pid, tmpdir)
+    time.sleep(1)
```
Comment (LeonidVas) [inline test/integration/running/test_running.py:220]: Why you do a sleep? Why 1 and not 10?
I think we can check this event (end of work) by PID.

**Comment by LeonidVas** [comment]
Do a rebase on the master.

File: test/integration/running/test_running.py [hunk 15]
```diff
@@ -139,3 +140,91 @@ def test_logrotate(tt_cmd, tmpdir):
...
+
+    # Create test config with restart_on_failure false.
+    with open(configPath, 'w') as f:
+        f.write("tt:\n  app:\n    restart_on_failure: false\n    log_dir: "+"str(tmpdir)+\n    log_maxsize: 10    \n    log_maxage: 1\n")
+    
+    # Wait for child process of instance to start.
+    flag = 0
+    while flag == 0:
+        try:
+            with open(logPath, "r") as f:
```
Comment (LeonidVas) [inline test/integration/running/test_running.py:209]: Please, don't use one letter names.

File: test/integration/running/test_running.py [hunk 16]
```diff
@@ -139,3 +140,91 @@ def test_logrotate(tt_cmd, tmpdir):
...
+    assert status_rc == 0
+    assert re.search(r"RUNNING. PID: \d+.", status_out)
+
+    res = [int(i) for i in status_out if i.isdigit()]
+    pid = ""
+    for digit in res:
+        pid += str(digit)
+    print("Pid: ", pid)
+
+    # Wait for child process of instance to start.
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply test/integration/running/test_running.py:234]: added

File: test/integration/running/test_running.py [hunk 17]
```diff
@@ -139,3 +140,91 @@ def test_logrotate(tt_cmd, tmpdir):
...
+    logPath = os.path.join(tmpdir, inst_name + ".log")
+    flag = 0
+    while flag == 0:
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply test/integration/running/test_running.py]: added timeout

File: test/utils.py [hunk 18]
```diff
@@ -79,3 +80,22 @@ def wait_file(dir_name, file_pattern, exclude_list, timeout_sec=1):
         iter_count = iter_count + 1
 
     return ""
+
+def kill_child_process(pid, cwd):
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply test/utils.py]: added

[the remark this answers is not in the document]
Comment (better0fdead) [reply test/utils.py]: deleted

File: test/utils.py [hunk 19]
```diff
@@ -79,3 +80,22 @@ def wait_file(dir_name, file_pattern, exclude_list, timeout_sec=1):
...
+    parent = psutil.Process(int(pid))
+    ch_pid_list = []
+    while len(ch_pid_list) < 1:
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply test/utils.py]: changed

File: test/utils.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (better0fdead) [reply test/utils.py]: changed

File: test/integration/running/test_running.py (hunk 12 above)
Comment (better0fdead) [reply test/integration/running/test_running.py:249]: added descrption

File: test/integration/running/test_running.py (hunk 13 above)
Comment (better0fdead) [reply test/integration/running/test_running.py]: moved

File: test/integration/running/test_running.py (hunk 14 above)
Comment (better0fdead) [reply test/integration/running/test_running.py]: removed

File: test/integration/running/test_running.py (hunk 15 above)
Comment (better0fdead) [reply test/integration/running/test_running.py]: changed

File: test/integration/running/test_running.py [hunk 20]
```diff
@@ -139,3 +140,91 @@ def test_logrotate(tt_cmd, tmpdir):
...
+    start_output = instance_process.stdout.readline()
+    assert re.search(r"Starting an instance", start_output)
+
+    # Get pid of instance.
+    status_cmd = [tt_cmd, "status", inst_name]
+    status_rc, status_out = run_command_and_get_output(status_cmd, cwd=tmpdir)
+    assert status_rc == 0
+    assert re.search(r"RUNNING. PID: \d+.", status_out)
+
+    res = [int(i) for i in status_out if i.isdigit()]
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply test/integration/running/test_running.py]: changed

File: test/integration/running/test_running.py [hunk 21]
```diff
@@ -139,3 +140,91 @@ def test_logrotate(tt_cmd, tmpdir):
...
+        start_cmd,
+        cwd=tmpdir,
+        stderr=subprocess.STDOUT,
+        stdout=subprocess.PIPE,
+        text=True
+    )
+    start_output = instance_process.stdout.readline()
+    assert re.search(r"Starting an instance", start_output)
+
+    # Get pid of instance.
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply test/integration/running/test_running.py:224]: added

File: test/utils.py [hunk 22]
```diff
@@ -79,3 +80,22 @@ def wait_file(dir_name, file_pattern, exclude_list, timeout_sec=1):
...
+        ch_pid_list = parent.children()
+    ch_pid = ch_pid_list[0].pid
+    print("chPid: ", ch_pid)
+    ch_pid = int(ch_pid)
+    kill_cmd = ["kill", str(ch_pid)]
+    kill_process = subprocess.Popen(
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply test/utils.py]: changed

**Pull request merged**